packages/next/src/export/index.ts

```diff
@@ -148,746 +146,761 @@ export class ExportError extends Error {
-  span: Span
-): Promise<void> {
-  const nextExportSpan = span.traceChild('next-export')
+function setupWorkers(
+  options: ExportAppOptions,
+  nextConfig: NextConfigComplete
+): ExportWorkers {
+  if (options.exportPageWorker) {
+    return {
+      pages: options.exportPageWorker,
+      app: options.exportAppPageWorker,
+      end: options.endWorker || (() => Promise.resolve()),
+    }
+  }
 
-  return nextExportSpan.traceAsyncFn(async () => {
-    dir = resolve(dir)
-
-    // attempt to load global env values so they are available in next.config.js
-    nextExportSpan
-      .traceChild('load-dotenv')
-      .traceFn(() => loadEnvConfig(dir, false, Log))
-
-    const nextConfig =
-      options.nextConfig ||
-      (await nextExportSpan
-        .traceChild('load-next-config')
-        .traceAsyncFn(() => loadConfig(PHASE_EXPORT, dir)))
-
-    const threads = options.threads || nextConfig.experimental.cpus
-    const distDir = join(dir, nextConfig.distDir)
-    const isExportOutput = nextConfig.output === 'export'
-
-    // Running 'next export'
-    if (options.isInvokedFromCli) {
-      if (isExportOutput) {
-        if (options.hasOutdirFromCli) {
-          throw new ExportError(
-            '"next export -o <dir>" cannot be used when "output: export" is configured in next.config.js. Instead add "distDir" in next.config.js https://nextjs.org/docs/advanced-features/static-html-export'
-          )
-        }
+  const threads = options.threads || nextConfig.experimental.cpus
+  if (!options.silent && !options.buildExport) {
+    Log.info(`Launching ${threads} workers`)
+  }
+
+  const timeout = nextConfig?.staticPageGenerationTimeout || 0
+
+  let infoPrinted = false
+
+  const worker = new Worker(require.resolve('./worker'), {
+    timeout: timeout * 1000,
+    onRestart: (_method, [{ path }], attempts) => {
+      if (attempts >= 3) {
+        throw new ExportError(
+          `Static page generation for ${path} is still timing out after 3 attempts. See more info here https://nextjs.org/docs/messages/static-page-generation-timeout`
+        )
+      }
+      Log.warn(
+        `Restarted static page generation for ${path} because it took more than ${timeout} seconds`
+      )
+      if (!infoPrinted) {
         Log.warn(
-          '"next export" is no longer needed when "output: export" is configured in next.config.js https://nextjs.org/docs/advanced-features/static-html-export'
+          'See more info here https://nextjs.org/docs/messages/static-page-generation-timeout'
         )
-        return
+        infoPrinted = true
       }
-      if (existsSync(join(distDir, 'server', 'app'))) {
+    },
+    maxRetries: 0,
+    numWorkers: threads,
+    enableWorkerThreads: nextConfig.experimental.workerThreads,
+    exposedMethods: ['default'],
+  })
+
+  return {
+    pages: (worker as any).default as ExportWorker,
+    end: async () => {
+      await worker.end()
+    },
+  }
+}
+
+export async function exportAppImpl(
```

Comment:
While I don't agree with this naming, I feel like we should get into the habit of doing this more often. It's very unpleasent to have to work with a function that does:

```
function someThing(span) {
  return span.traceFn(() => {
    // ...
  })
}
```

Instead layering this with different functions is much nicer. Possibly we should explore using decorators in the future for this type of work?